### PR TITLE
Fix curried type annotations

### LINKED ISSUE #156: Curried functions don't play nicely with type hinting
### What was wrong?

Now that `eth-typing` is properly exposing types for functions and `mypy` is better at using those, this code throws errors.

```python
from_hex_to_bytes = hexstr_if_str(to_bytes)
# error: Too few arguments for "hexstr_if_str"
```

### How can it be fixed?

There may be a better way but the following seems to work.

Use [`typing.overload`](https://docs.python.org/3/library/typing.html#typing.overload) to manually declare the various allowed signatures and then define the curried version. 

```python
T = TypeVar("T")


@overload
def hexstr_if_str() -> Callable[[Callable[..., T], Union[bytes, int, str]], T]:
    pass


@overload
def hexstr_if_str(to_type: Callable[..., T]) -> Callable[[Union[bytes, int, str]], T]:
    pass


@overload
def hexstr_if_str(
    to_type: Callable[..., T], hexstr_or_primitive: Union[bytes, int, str]
) -> T:
    pass


def hexstr_if_str(
    to_type: Callable[..., T] = None, hexstr_or_primitive: Union[bytes, int, str] = None
) -> T:
    pass


hexstr_if_str = curry(_hexstr_if_str)
```

### PR DESCRIPTION
**Pull request by carver**
### What was wrong?

Fixes #156 

### How was it fixed?

Add `@overload` definitions in the `curried` module.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/eth-utils/blob/master/newsfragments/README.md)

- [x] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by carver** [comment]
I checked this against py-evm, which now passes its mypy check.

File: eth_utils/curried/__init__.py [hunk 1]
```diff
@@ -82,20 +82,122 @@
...
 from_wei = curry(from_wei)
 get_logger = curry(get_logger)
-hexstr_if_str = curry(hexstr_if_str)
+hexstr_if_str = curry(non_curried_hexstr_if_str)
 is_same_address = curry(is_same_address)
 text_if_str = curry(text_if_str)
 to_wei = curry(to_wei)
 clamp = curry(clamp)
 
+del Any
```
Comment (pipermerriam) [inline eth_utils/curried/__init__.py:191]: Curious at using this approach as opposed to `__all__`?  At minimum a comment above these `del` statements explaining why they are here would :smile:

Comment (carver) [reply eth_utils/curried/__init__.py:191]: Didn't put too much thought into it. Basically something like:
- There was already a `del` there
- The list of things to delete is shorter than the list of things to import

I'll try it out and see how it it looks the other way.

Comment (carver) [reply eth_utils/curried/__init__.py:191]: Hm, `__all__` doesn't actually remove the others from the import list, it only affects `from eth_utils.curried import *`, which we don't use anyway.

When I use `__all__`, this happens:
```py
>>> import eth_utils.curried
>>> eth_utils.curried.Any
typing.Any
>>> from eth_utils.curried import *
>>> to_hex
<function eth_utils.conversions.to_hex(primitive:Union[bytes, int]=None, hexstr:<function NewType.<locals>.new_type at 0x7fcf40edb158>=None, text:str=None) -> <function NewType.<locals>.new_type at 0x7fcf40edb158>>
>>> Any
NameError: name 'Any' is not defined
>>> from eth_utils.curried import Any
>>> Any
typing.Any

**Comment by carver** [comment]
> I sort of expected to see new tests in https://github.com/ethereum/eth-utils/tree/c5595dda76e3a0150a36a2b4cdfb4594e3631444/tests/type-checks but I don't have any specific tests to suggest...

Yeah, the main issue seemed to be checking that all the methods are importable, so I added that check. (and confirmed that it fails before the change, although I did that check with a more recent mypy)

File: eth_utils/curried/__init__.py (hunk 1 above)
Comment (pipermerriam) [reply eth_utils/curried/__init__.py:191]: Yeah, I forgot that `__all__` only effects `*`.  Dumb python...

**Pull request merged**